server/cluster/shard.go

```diff
@@ -21,3 +24,57 @@ type ShardTablesWithRole struct {
 	tables    []*Table
 	version   uint64
 }
+
+// fsm const
+const (
+	EventTransferLeader           = "TransferLeader"
+	EventTransferToFollowerFailed = "TransferToFollowerFailed"
+	EventTransferFollower         = "TransferFollower"
+	EventTransferToLeaderFailed   = "TransferToLeaderFailed"
+	EventTransferFollowerStart    = "TransferFollowerStart"
+	EventTransferLeaderStart      = "TransferLeaderStart"
+	StateLeader                   = "LEADER"
+	StateFollower                 = "FOLLOWER"
+	StatePendingLeader            = "PENDING_LEADER"
+	StatePendingFollower          = "PENDING_FOLLOWER"
+)
+
+// NewFSM /**
+/**
+```
+┌────┐                   ┌────┐
+│ RW ├─────────┐         │ R  ├─────────┐
+├────┘         │         ├────┘         │
+│    Leader    ◀─────────│PendingLeader │
+│              │         │              │
+└───────┬──▲───┘         └───────▲─┬────┘
+        │  │                     │ │
+┌────┐  │  │             ┌────┐  │ │
+│ R  ├──▼──┴───┐         │ R  ├──┴─▼────┐
+├────┘         │         ├────┘         │
+│   Pending    ├─────────▶   Follower   │
+│   Follower   │         │              │
+└──────────────┘         └──────────────┘
+```
+*/
+func NewFSM(role clusterpb.ShardRole) *fsm.FSM {
+	fsm := fsm.NewFSM(
+		StateFollower,
+		fsm.Events{
+			{Name: EventTransferLeader, Src: []string{StatePendingLeader}, Dst: StateLeader},
+			{Name: EventTransferToFollowerFailed, Src: []string{StatePendingFollower}, Dst: StateLeader},
+			{Name: EventTransferFollower, Src: []string{StatePendingFollower}, Dst: StateFollower},
+			{Name: EventTransferToLeaderFailed, Src: []string{StatePendingLeader}, Dst: StateFollower},
+			{Name: EventTransferFollowerStart, Src: []string{StateLeader}, Dst: StatePendingFollower},
+			{Name: EventTransferLeaderStart, Src: []string{StateFollower}, Dst: StatePendingLeader},
+		},
```

Comment:
The state transitions seem unchangeable, so shall we define it as a constant variable shared by all the created `FSM`.